src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInEnumerationCheck.java

```diff
@@ -0,0 +1,110 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2019 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks if unnecessary semicolon is in enum definitions.
+ * Semicolon is not needed if enum body is empty.
+ * The check will flag the following with warnings:
+ * </p>
+ * <pre>
+ * enum One {
+ *     A,B; // violation
+ * }
+ * enum Two {
+ *     A,B,; // violation
+ * }
+ * enum Three {
+ *     A,B(); // violation
+ * }
+ * enum Four {
+ *     A,B{}; // violation
+ * }
+ * enum NoSemicolon {
+ *     A,B // OK
+ * }
+ * enum Normal {
+ *     A,
+ *     B
+ *     ; // required ";", no violation
+ *     Normal(){}
+ * }
+ * </pre>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;UnnecessarySemicolonInEnumeration&quot;/&gt;
+ * </pre>
+ *
+ * @since 8.22
+ */
+@StatelessCheck
+public final class UnnecessarySemicolonInEnumerationCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_SEMI = "unnecessary.semicolon";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.ENUM_CONSTANT_DEF,
```

Comment:
Why is violation on `ENUM_CONSTANT_DEF` instead of `ENUM_DEF`?
Aren't you printing multiple duplicate violations if there are multiple enum definitions inside. (This is still regardless that we ignore duplicate violations)
Also, aren't you missing on printing a violation for a completely empty enum?
Example:
````
public enum Example { ; }
````
If it is valid, please add it as another test case too.

Reply:
Yeap, I missed that case.

I updated code and test cases.